### ISSUE TITLE
HDFS Connector caught in recovery loop 

### ISSUE THREAD

**Issue by pederpansen** [confluentinc/kafka-connect-hdfs#264]
We have a number of HDFS Connectors running in a Kafka Connect 3.2.2 runtime.

While the connectors were working correctly initially, they are now caugh in a "recovery loop". Example log output (grepped for a sample partition):

`[2017-12-15 11:03:01,682] INFO Started recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)
[2017-12-15 11:03:14,033] INFO Finished recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)
[2017-12-15 11:16:40,685] INFO Started recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)
[2017-12-15 11:16:53,452] INFO Finished recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)
[2017-12-15 11:29:44,131] INFO Started recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)
[2017-12-15 11:29:56,256] INFO Finished recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)
[2017-12-15 11:42:02,921] INFO Started recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)
[2017-12-15 11:42:15,230] INFO Finished recovery for topic partition mytopic-11 (io.confluent.connect.hdfs.TopicPartitionWriter)`

This keeps on going on and on, for a large number of topic/partitions. 

Deleting the respective HDFS folders for topics (containing .parquet data files) and logs (containing WAL files) lets the partition recover. 

However, this is no solution available in production environments, as it causes data loss.

**Comment by alexandrfox** [comment]
@pederpansen if along those lines you also see new assignments for partitions in the log that would mean your consumer group is continuously rebalancing. Try to tune your session timeout and max poll interval settings for consumer, e.g.:
`
consumer.session.timeout.ms=10000
consumer.heartbeat.interval.ms=3000
consumer.request.timeout.ms=1810000
consumer.max.poll.interval.ms=1800000
`

**Comment by pederpansen** [comment]
This problem persists in Confluent 4.1.

@alexandrfox There are no new assignments in the log.

**Comment by pederpansen** [comment]
I believe I found the cause for my problem.

I was not aware that there are actually multiple directories in HDFS where the connector stores data or metadata:

- `<configured_root>/topics/<my-topic>`
- `<configured_root>/logs/<my-topic>`
- `<configured_root>/topics/+tmp/<my-topic>`

I had previously deleted the first two folders, but not the last. There were old files in the `+tmp` folder which apparently caused the recovery to fail. 

I'm not sure what info is stored in those files, but it seemed to confuse the connector catching up to the latest written offsets (possibly, those files had offsets in them that were not existing in the topic anymore because of the configured retention time).

I stopped the connector, deleted _all three folders_ from HDFS, started the connector and the recovery went fine.

**Comment by Inylschek** [comment]
Hi, I disagree that this is a solution. It works, but then you have the problem of having data split between two different sets of directories.

It doesn't actually 'fix' the problem

**Comment by Inylschek** [comment]
> @pederpansen if along those lines you also see new assignments for partitions in the log that would mean your consumer group is continuously rebalancing. Try to tune your session timeout and max poll interval settings for consumer, e.g.:
> 
> ```
> consumer.session.timeout.ms=10000
> consumer.heartbeat.interval.ms=3000
> consumer.request.timeout.ms=1810000
> consumer.max.poll.interval.ms=1800000
> ```

Where does one make these changes? I know these configurations from the documentation, but my understanding is that consumer properties are altered in the application doing the consuming, and I don't see these as options for the Hdfs Sink Connector, or even within its code base

**Comment by koen-dejonghe** [comment]
'I stopped the connector, deleted all three folders from HDFS, started the connector and the recovery went fine.'
How is this a solution? Deleting your data is a solution?

**Comment by pederpansen** [comment]
Hi @koen-dejonghe, to answer your (I guess rhetorical) question

> How is this a solution? Deleting your data is a solution?

That depends. If your retention time is long enough (which in my case it was), it might be totally fine. Of course, this is more of a crude workaround.

I would suggest to try a recent version of the connector. We suspected the issue was related to https://github.com/confluentinc/kafka-connect-hdfs/issues/53, which was fixed in the meantime.
